File: 2019-2020/pset2/vigenere.c
/**
 * vigenere.c
 *
 * Computer Science 50
 * Problem Set 2
 *
 * Encrypts messages using Vigenère’s cipher
 *
 * 2019:
 * https://lab.cs50.io/cs50/labs/2019/x/vigenere/
 *
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <cs50.h>

// Prototypes
char charCaesarEncrypt(char, int);
string stringVigenereEncrypt(string, string);

// MAIN
int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    
    string keyWord = argv[1];
    
    char c;
    for (int i = 0; (c = keyWord[i]) != '\0'; ++i)
    {
        if (!isalpha(c))
        {
            printf("Usage: ./vigenere keyword\n");
            return 1;
        }    
    }
    
    string input = get_string("plaintext: ");
            
    printf("ciphertext: %s\n", stringVigenereEncrypt(input, keyWord));
    return 0;
}

// Functions
// -------------------------------------------------------
char charCaesarEncrypt(char c, int key)
{
    int shiftedC = c + key % 26;

    if (islower(c))
    {
        return 'a' + (shiftedC - 'a') % 26;
    }
    if (isupper(c))
    {
        return 'A' + (shiftedC - 'A') % 26;
    }

    return c;
}

// ----------------------------------------------------------
string stringVigenereEncrypt(string input, string keyWord)
{
    int len = strlen(input);
    int lenKeyWord = strlen(keyWord);
    string output = (string) malloc(len * sizeof(char) + 1);
       
    for (int i = 0, j = 0; i < len; ++i)
    {
        if (j == lenKeyWord)
        {
            j = 0;
        }
        output[i] = charCaesarEncrypt(input[i], tolower(keyWord[j]) - 'a');
        
        if (isalpha(input[i]))
        {
            ++j;
        }
    }
    output[len] = '\0'; 
    return output;
}

File: old/2019-2020/pset6/vigenere.py
# ------------------------------------------------------------------------------
# vigenere.py
#
# Computer Science 50
# Problem Set 6
#
# Encrypts messages using Vigenère’s cipher
#
# 2019:
# https://docs.cs50.net/2019/x/psets/6/sentimental/vigenere/vigenere.html
# ------------------------------------------------------------------------------

from sys import argv, exit


def main():
    if len(argv) != 2 or not argv[1].isalpha():
        print("Usage: python vigenere.py k")
        exit(1)

    s = input("plaintext:  ")
    print("ciphertext: {0}".format(vigenere_encrypt(s, argv[1])))


def caesar_symbol_encrypt(sym, shift):
    if 'a' <= sym <= 'z':
        sym = chr(ord('a') + (ord(sym) + shift - ord('a')) % 26)
    elif 'A' <= sym <= 'Z':
        sym = chr(ord('A') + (ord(sym) + shift - ord('A')) % 26)

    return sym


def vigenere_encrypt(str, keyword):
    keyword = keyword.lower()
    tmp_list = []
    k = 0
    kyeword_len = len(keyword)
    ord_a = ord('a')
    for c in str:
        if not c.isalpha():
            tmp_list.append(c)
            continue
        key_symbol = keyword[k % kyeword_len]
        tmp_list.append(caesar_symbol_encrypt(c, ord(key_symbol) - ord_a))
        k += 1
    return ''.join(tmp_list)


if __name__ == "__main__":
    main()
